cmd/gui/gui.go

```diff
@@ -0,0 +1,137 @@
+package gui
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"path/filepath"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/cmd/agent/common"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	log "github.com/cihub/seelog"
+	"github.com/gorilla/mux"
+	"github.com/urfave/negroni"
+)
+
+var (
+	apiKey   string
+	listener net.Listener
+)
+
+// Payload struct is for the JSON messages received from a client POST request
+type Payload struct {
+	Config string `json:"config"`
+	Email  string `json:"email"`
+	CaseID string `json:"caseID"`
+}
+
+// StopGUIServer closes the connection to the HTTP server
+func StopGUIServer() {
+	if listener != nil {
+		listener.Close()
+	}
+}
+
+// StartGUIServer creates the router and starts the HTTP server
+func StartGUIServer() error {
+	apiKey = config.Datadog.GetString("api_key")
+	port := config.Datadog.GetString("GUI_port")
+
+	if config.Datadog.GetString("GUI_port") == "-1" {
+		log.Infof("Port -1 specified: not starting the GUI server.")
+		return nil
+	}
+
+	// Instantiate the gorilla/mux router
+	router := mux.NewRouter()
+
+	// Serve the (secured) index page on the default endpoint
+	router.Handle("/", accessAuth(http.FileServer(http.Dir(filepath.Join(common.GetViewPath(), "private")))))
+
+	// Mount our public filesystem at the view/{path} route
+	router.PathPrefix("/view/").Handler(http.StripPrefix("/view/", http.FileServer(http.Dir(filepath.Join(common.GetViewPath(), "public")))))
+
+	// Mount our secured filesystem at the private/{path} route
+	router.PathPrefix("/private/").Handler(http.StripPrefix("/private/", accessAuth(http.FileServer(http.Dir(filepath.Join(common.GetViewPath(), "private"))))))
+
+	// Set up handlers for the API
+	agentRouter := mux.NewRouter().PathPrefix("/agent").Subrouter().StrictSlash(true)
+	agentHandler(agentRouter)
+	checkRouter := mux.NewRouter().PathPrefix("/checks").Subrouter().StrictSlash(true)
+	checkHandler(checkRouter)
+
+	// Add authorization middleware to all the API endpoints
+	router.PathPrefix("/agent").Handler(negroni.New(negroni.HandlerFunc(authorize), negroni.Wrap(agentRouter)))
+	router.PathPrefix("/checks").Handler(negroni.New(negroni.HandlerFunc(authorize), negroni.Wrap(checkRouter)))
+
+	listener, e := net.Listen("tcp", ":"+port)
```

Comment:
As @derekwbrown already pointed out, we cannot really ship the GUI like this. I see the value of accessing it from different networks but at the bare minimum make this configurable, defaulting to `localhost`.

Reply:
For the moment, IMO, don't even make it configurable, only `localhost`.  Otherwise we open a can of worms we don't want to open WRT supporting remote administration via this gui.